### PR TITLE
DS-4096: updating owning collections

### PR DESCRIPTION
**Pull request by KevinVdV**
In this PR we have created support for moving an item between collections.
A new rest controller (ItemOwningCollectionUpdateRestController) was added to support the behavior which works using PUT.

Jira link: https://jira.duraspace.org/browse/DS-4096
RestContract PR link: DSpace/Rest7Contract#34

### CODE REVIEW

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    AuthorizeService authorizeService;
+
+    @Autowired
+    CollectionConverter converter;
+
+    @RequestMapping(method = RequestMethod.PUT)
+    @PreAuthorize("hasPermission(#itemUuid, 'ITEM','WRITE') && hasPermission(#targetUuid,'COLLECTION','ADD')")
+    @PostAuthorize("returnObject != null")
+    public CollectionRest move(@PathVariable UUID itemUuid, HttpServletResponse response,
+                               HttpServletRequest request, @RequestParam(name = "collection") UUID targetUuid)
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java]: This doesn't look to align with the REST Contract (see https://github.com/DSpace/Rest7Contract/pull/34).  The REST Contract stats that the `PUT` request should pass in the Collection's UUID in the *body* of the request (as a `text/uri-list`).   This method looks to be expecting the Collection UUID as a request parameter (`?collection=[:uuid]`)

**Review by tdonohue — CHANGES_REQUESTED** [review]
@KevinVdV : I reviewed this today.  The overall code looks reasonable, but the URI implementation here doesn't match what was decided upon in the REST Contract.  So, this code (and the Integration Test) should be updated to align with the REST Contract, see https://github.com/DSpace/Rest7Contract/pull/34

**Review by samuelcambien — APPROVED** [review]
method has been changed accordingly

**Comment by KevinVdV** [comment]
This PR is now ready for review, although it should not be merged before: https://github.com/DSpace/DSpace/pull/2403 because this PR includes those changes as well to handle some exceptions.

**Review by benbosman — APPROVED** [review]
I've reviewed this PR in detail as requested in last week's meeting.
I don't see any issues, and the ITs are present as well
The issues initially reported and requiring changes were applied

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java (hunk 1 above)
Comment (benbosman) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java]: This has been resolved

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/core/items/" +
+        "{itemUuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12" +
+        "}}/owningCollection")
+public class ItemOwningCollectionUpdateRestController {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java]: This Controller is missing JavaDocs, as are its public methods.

**Review by tdonohue — COMMENTED** [review]
Overall, this code looks reasonable. I haven't had a chance to test it yet. But, I did notice that it is currently missing Javadocs.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java (hunk 2 above)
Comment (YanaDePauw) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java]: Added javadocs to the ItemOwningCollectionUpdateRestController

**Review by tdonohue — APPROVED** [review]
👍 Rereviewed today, and this now looks good! Thanks for adding the missing Javadocs @YanaDePauw !

**Pull request merged**